REGRESSION (273804@main): [ macOS wk1 Debug ] http/tests/media/video-throttled-load-metadata.html is a constant crash

jyavenard opened the pull request:
#### e70021fd8743d5ae2174bbf43aaa707dc505b02b
<pre>
REGRESSION (273804@main): [ macOS wk1 Debug ] http/tests/media/video-throttled-load-metadata.html is a constant crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=268480">https://bugs.webkit.org/show_bug.cgi?id=268480</a>
<a href="https://rdar.apple.com/122025990">rdar://122025990</a>

Reviewed by Youenn Fablet.

273804@main incorrectly removed the delayed task removal to perform the operation immediately.
However, this method is called while iterating the HashTable from which we are removing the task causing
the iterator to become invalid.
Re-introduce the delayed operation, add comment in code.

Covered by existing tests.

* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::RangeResponseGenerator::giveResponseToTaskIfBytesInRangeReceived):

Canonical link: <a href="https://commits.webkit.org/273892@main">https://commits.webkit.org/273892@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cad1c5686bc1d11e1466a08701d56ebcf5e14567

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37038 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15949 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39342 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39620 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33099 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38328 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13082 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31614 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37600 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13453 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32651 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11721 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11743 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32951 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40876 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33556 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33460 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37635 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12047 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9851 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35769 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13691 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8379 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12421 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12955 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf reviewed (approved):
LGTM.
Quick question with the model though: why do we need to remove it once consumed?
Could we remove it from the map before calling giveResponseToTasksWithFinishedRanges? And pass ownership to giveResponseToTasksWithFinishedRanges?

jyavenard committed:
REGRESSION (273804@main): [ macOS wk1 Debug ] http/tests/media/video-throttled-load-metadata.html is a constant crash
https://bugs.webkit.org/show_bug.cgi?id=268480
rdar://122025990

Reviewed by Youenn Fablet.

273804@main incorrectly removed the delayed task removal to perform the operation immediately.
However, this method is called while iterating the HashTable from which we are removing the task causing
the iterator to become invalid.
Re-introduce the delayed operation, add comment in code.

Covered by existing tests.

* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::RangeResponseGenerator::giveResponseToTaskIfBytesInRangeReceived):

Canonical link: https://commits.webkit.org/273892@main
